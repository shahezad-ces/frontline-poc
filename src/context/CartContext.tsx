"use client";
import { createContext, useContext, useEffect, useState } from "react";
import { getLocalStorage, setLocalStorage } from "@frontline/utils";
import type { CartContextType, CartItem, Product } from "@frontline/types";

const CartContext = createContext<CartContextType | undefined>(undefined);

export const CartProvider = ({ children }: { children: React.ReactNode }) => {
  const [cart, setCart] = useState<CartItem[]>([]);
  const [quantity, setQuantity] = useState(1);

  // Load from localStorage
  useEffect(() => {
    const storedCart = getLocalStorage<CartItem[]>("cart");
    if (storedCart) setCart(storedCart);
  }, []);

  // Save to localStorage
  useEffect(() => {
    setLocalStorage("cart", cart);
  }, [cart]);

  const addToCart = (product: Product) => {
    setCart((prev) => {
      const existing = prev.find((item) => item.id === product.id);
      if (existing) {
        return prev.map((item) =>
          item.id === product.id
            ? { ...item, quantity: item.quantity + quantity }
            : item
        );
      }
      return [...prev, { ...product, quantity }];
    });
    setQuantity(1);
  };

  const updateQuantity = (quantity: number) => {
    if (quantity < 1) return; // Prevent zero or negative
    setQuantity(quantity);
  };

  const removeFromCart = (id: number) => {
    setCart((prev) => prev.filter((item) => item.id !== id));
  };

  const clearCart = () => setCart([]);

  const totalItems = cart.reduce((sum, item) => sum + item.quantity, 0);
  const totalPrice = cart.reduce(
    (sum, item) => sum + item.price * item.quantity,
    0
  );

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        updateQuantity,
        removeFromCart,
        clearCart,
        totalItems,
        totalPrice,
        quantity,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  const context = useContext(CartContext);
  if (!context) throw new Error("useCart must be used within CartProvider");
  return context;
};
