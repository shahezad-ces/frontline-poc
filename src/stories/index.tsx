import Pagination from "./Pagination/Pagination";
import Button from "./Button/Button";
import Image from "./Image/Image";
import QuantitySelector from "./QuantitySelector/QuantitySelector";
import Toast from "./Toast/Toast";

export { Pagination, Button, Image, QuantitySelector, Toast };
