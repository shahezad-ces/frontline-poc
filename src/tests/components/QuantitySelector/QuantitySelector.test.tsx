import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { useCart } from "@frontline/context/CartContext";
import { QuantitySelector } from "@frontline/stories";

// Mock useCart hook
jest.mock("@frontline/context/CartContext", () => ({
  useCart: jest.fn(),
}));

const mockUseCart = useCart as jest.Mock;

describe("QuantitySelector Component", () => {
  const updateQuantityMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    mockUseCart.mockReturnValue({
      quantity: 5,
      updateQuantity: updateQuantityMock,
    });
  });

  it("renders with default quantity", () => {
    render(<QuantitySelector />);
    expect(screen.getByRole("spinbutton")).toHaveValue(5);
  });

  it("increases quantity when plus button is clicked", () => {
    render(<QuantitySelector />);
    const plusButton = screen.getAllByRole("button")[1];
    fireEvent.click(plusButton);
    expect(updateQuantityMock).toHaveBeenCalledWith(6);
  });

  it("decreases quantity when minus button is clicked", () => {
    render(<QuantitySelector />);
    const minusButton = screen.getAllByRole("button")[0];
    fireEvent.click(minusButton);
    expect(updateQuantityMock).toHaveBeenCalledWith(4);
  });

  it("prevents quantity from going below min", () => {
    mockUseCart.mockReturnValue({
      quantity: 1,
      updateQuantity: updateQuantityMock,
    });
    render(<QuantitySelector min={1} />);
    const minusButton = screen.getAllByRole("button")[0];
    fireEvent.click(minusButton);
    expect(updateQuantityMock).not.toHaveBeenCalled();
  });

  it("prevents quantity from going above max", () => {
    mockUseCart.mockReturnValue({
      quantity: 99,
      updateQuantity: updateQuantityMock,
    });
    render(<QuantitySelector max={99} />);
    const plusButton = screen.getAllByRole("button")[1];
    fireEvent.click(plusButton);
    expect(updateQuantityMock).not.toHaveBeenCalled();
  });

  it("updates quantity when input value changes", () => {
    render(<QuantitySelector />);
    const input = screen.getByRole("spinbutton");
    fireEvent.change(input, { target: { value: "10" } });
    expect(updateQuantityMock).toHaveBeenCalledWith(10);
  });

  it("ignores invalid input (negative or too large)", () => {
    render(<QuantitySelector min={1} max={10} />);
    const input = screen.getByRole("spinbutton");

    fireEvent.change(input, { target: { value: "-5" } });
    expect(updateQuantityMock).not.toHaveBeenCalled();

    fireEvent.change(input, { target: { value: "100" } });
    expect(updateQuantityMock).not.toHaveBeenCalled();
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<QuantitySelector />);
    expect(asFragment()).toMatchSnapshot();
  });
});
