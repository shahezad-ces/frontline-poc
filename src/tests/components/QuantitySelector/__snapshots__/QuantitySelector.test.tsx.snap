// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`QuantitySelector Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex items-center border border-gray-200 rounded-lg w-full md:w-50"
  >
    <button
      class="w-20 h-10 p-2 flex items-center justify-center text-blue-600 text-2xl border-r border-gray-200 rounded-l-lg hover:bg-gray-200 cursor-pointer"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6 disabled:text-gray-200"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4 10a.75.75 0 0 1 .75-.75h10.5a.75.75 0 0 1 0 1.5H4.75A.75.75 0 0 1 4 10Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <input
      class="w-full text-center outline-none text-lg"
      type="number"
      value="5"
    />
    <button
      class="w-20 h-10 p-2 flex items-center justify-center text-blue-600 text-2xl border-l border-gray-200 rounded-r-lg hover:bg-gray-200 cursor-pointer"
    >
      <svg
        aria-hidden="true"
        class="w-6 h-6"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10.75 4.75a.75.75 0 0 0-1.5 0v4.5h-4.5a.75.75 0 0 0 0 1.5h4.5v4.5a.75.75 0 0 0 1.5 0v-4.5h4.5a.75.75 0 0 0 0-1.5h-4.5v-4.5Z"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
